read_html(): rewinding [wip]

erinzm opened the pull request:
- [x] closes #17975
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jreback commented:
tests!

liamim commented:
@jreback yep, they're coming :smile:

liamim commented:
Unfortunately, non-seekable io objects typically still have a stub `.seek()` implementation which raises `io.UnsupportedOperation`. I'll trap that instead.

erinzm committed:
io.html.read_html(): rewind seekable io objects when parsers fail

If lxml has read to the end of a file and then errored,
bs4/html5lib won't rewind it before trying to parse again,
and will throw a `ValueError: No text parsed from document`.

This patch fixes this issue, by rewinding the file object
when a parser fails. If the object was IO-ish but not seekable,
we throw an error notifying the user and asking them to try
a different flavor.

jreback commented on pandas/io/html.py:
```diff
@@ -742,6 +742,19 @@ def _parse(flavor, io, match, attrs, encoding, **kwargs):
+            # and try to rewind it before trying the next parser
+            if hasattr(io, 'seekable') and io.seekable():
+                io.seek(0)
+
```
make an if else

erinzm committed:
test: read_html() parse error interaction with unseekable IO

pep8speaks commented:
Hello @liamim! Thanks for updating the PR.



 - In the file [`pandas/tests/io/test_html.py`](https://github.com/pandas-dev/pandas/blob/b16fcaf80b19cf7878e3910043e24ea8a3e332c7/pandas/tests/io/test_html.py), following are the PEP8 issues :

> [Line 990:1](https://github.com/pandas-dev/pandas/blob/b16fcaf80b19cf7878e3910043e24ea8a3e332c7/pandas/tests/io/test_html.py#L990): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

liamim commented:
How should I test rewinding on file objects? Just add a malformed HTML file to `pandas/tests/io/data`?

jreback commented:
there are several examples in the issue
you can also create a class that is has a read and seek method that wraps a url

jschendel commented on pandas/io/html.py:
```diff
@@ -742,6 +742,18 @@ def _parse(flavor, io, match, attrs, encoding, **kwargs):
+                io.seek(0)
+            elif hasattr(io, 'seekable') and not io.seekable():
+                # if we couldn't rewind it, let the user know
+                raise ValueError('The favor {} failed to parse your input. '
```
favor -> flavor

erinzm committed:
test: read_html() rewinding after parse errors

erinzm committed:
doc: whatsnew for read_html() rewinding fixes

jreback commented:
thanks @liamim nice PR!
